loadgen/loadgen.cc

```diff
@@ -1289,33 +1189,87 @@ void FindPeakPerformanceMode(SystemUnderTest* sut, QuerySampleLibrary* qsl,
     detail("Starting FindPeakPerformance mode:");
   });
 
-  TestSettingsInternal found_settings =
-      BinarySearchToFindPeakPerformance<scenario>(sut, qsl, sequence_gen,
-                                                  settings);
-
-  // Print-out the peak performance test setting.
-  LogDetail([field = find_peak_performance::ToStringPerformanceField<scenario>(
-                 found_settings)](AsyncDetail& detail) {
-    detail("FindPeakPerformance: Found peak performance field: " + field);
+  // Check whether settings satisfy the performance constraints or
+  // not.
+  LogDetail([initial_field =
+                 find_peak_performance::ToStringPerformanceField<scenario>(
+                     settings)](AsyncDetail& detail) {
+    detail(
+        "FindPeakPerformance: Check validity of the base settings "
+        "field: " +
+        initial_field);
   });
 
-  // Must re-generate to avoid invalid set of samples by padded values of the
-  // MultiStream scenarios.
-  std::vector<loadgen::LoadableSampleSet> loadable_sets(
-      loadgen::GenerateLoadableSets(qsl, found_settings));
+  std::vector<loadgen::LoadableSampleSet> base_loadable_sets(
+      loadgen::GenerateLoadableSets(qsl, settings));
+  const LoadableSampleSet& base_performance_set = base_loadable_sets.front();
+  LoadSamplesToRam(qsl, base_performance_set.set);
 
-  // Use first loadable set as the performance set.
+  PerformanceResult base_pr(IssueQueries<scenario, TestMode::PerformanceOnly>(
+      sut, settings, base_performance_set, sequence_gen));
+  PerformanceSummary base_perf_summary{sut->Name(), settings,
+                                       std::move(base_pr)};
+
+  std::string msg;
+  if (!base_perf_summary.PerfConstraintsMet(&msg)) {
+    LogDetail([msg](AsyncDetail& detail) {
+      detail(
+          "FindPeakPerformance: Initial lower bound does not satisfy "
+          "performance constraints, msg: " +
+          msg);
+    });
+
+    sut->ReportLatencyResults(base_perf_summary.pr.latencies);
+
+    LogSummary(
+        [perf_summary = PerformanceSummary{sut->Name(), settings,
```

Comment:
To prevent the timing issue, make another struct here. Any better idea?

Reply:
What timing issue are you referring to?.